extensions/amp-video/0.1/amp-video.js

```diff
@@ -204,38 +243,131 @@ class AmpVideo extends AMP.BaseElement {
       return Promise.resolve();
     }
 
-    if (this.element.getAttribute('src')) {
-      assertHttpsUrl(this.element.getAttribute('src'), this.element);
-    }
+    const viewer = Services.viewerForDoc(this.getAmpDoc());
 
-    this.propagateAttributes(ATTRS_TO_PROPAGATE_ON_LAYOUT, this.video_,
+    this.propagateAttributes(ATTRS_TO_PROPAGATE_ON_LAYOUT,
+        dev().assertElement(this.video_),
         /* opt_removeMissingAttrs */ true);
 
-    this.getRealChildNodes().forEach(child => {
-      // Skip the video we already added to the element.
-      if (this.video_ === child) {
-        return;
-      }
-      if (child.getAttribute && child.getAttribute('src')) {
-        assertHttpsUrl(child.getAttribute('src'),
-            dev().assertElement(child));
+    if (viewer.getVisibilityState() == VisibilityState.PRERENDER) {
+      if (!this.element.hasAttribute('preload')) {
+        this.video_.setAttribute('preload', 'auto');
       }
-      this.video_.appendChild(child);
-    });
+      this.propagatePrerenderSources_();
+      viewer.whenFirstVisible().then(() => {
```

Comment:
Do you think this could introduce any side effects if the `source`s are added to the video if it is visible in the first viewport when the viewer becomes visible? I'm not very familiar with `source` tags. Does anything happen when new ones are added if a video is already visible?

Reply:
I only expect side-effects, unexpected behaviour when changing `source` after a video has started playing (which does not happen for us, even with autoplay). Based on my manual testing, seems to be working fine on Chrome and Safari.